source/byte_buf.c

```diff
@@ -53,26 +66,17 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
 }
 
 bool aws_byte_buf_is_valid(const struct aws_byte_buf *const buf) {
-    if (!buf) {
-        return false;
-    }
-    bool buffer_is_valid = (buf->buffer && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len));
-    bool capacity_is_valid = (buf->capacity > 0);
-    bool len_is_valid = (buf->len <= buf->capacity);
-    return capacity_is_valid && buffer_is_valid && len_is_valid;
+    return buf && ((buf->buffer == NULL && buf->capacity == 0 && buf->len == 0) ||
+                   (buf->capacity > 0 && buf->len <= buf->capacity && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len)));
 }
 
 bool aws_byte_cursor_is_valid(const struct aws_byte_cursor *cursor) {
```

Comment:
The new invariant assumes a 0 length `aws_byte_cursor` is valid always valid, any comments @justinboswell @bretambrose @danielsn ?